api/plugin.go

```diff
@@ -0,0 +1,390 @@
+// swagger:operation GET /plugin/:id/config plugin getPluginConfig
+//
+// Get YAML configuration for Configurer plugin.
+//
+// ---
+// consumes: [application/json]
+// produces: [application/x-yaml]
+// security: [clientTokenHeader: [], clientTokenQuery: [], basicAuth: []]
+// responses:
+//   200:
+//     description: Ok
+//     schema:
+//         type: object
+//         description: plugin configuration
+//   400:
+//     description: Bad Request
+//     schema:
+//         $ref: "#/definitions/Error"
+//   404:
+//     description: Not Found
+//     schema:
+//         $ref: "#/definitions/Error"
+//   500:
+//     description: Internal Server Error
+//     schema:
+//         $ref: "#/definitions/Error"
+func (c *PluginAPI) GetConfig(ctx *gin.Context) {
+	withID(ctx, "id", func(id uint) {
+		plugin := c.DB.GetPluginConfByID(id)
+		if plugin == nil {
+			ctx.AbortWithError(404, errors.New("unknown plugin"))
+			return
+		}
+		if aborted := pluginOwnerOrAbort(ctx, plugin); aborted {
+			return
+		}
+		instance, err := c.Manager.Instance(id)
+		if err != nil {
+			ctx.AbortWithError(404, errors.New("plugin instance not found"))
+			return
+		}
+
+		if aborted := supportOrAbort(ctx, instance, compat.Configurer); aborted {
+			return
+		}
+
+		ctx.Header("content-type", "application/x-yaml")
+		ctx.Writer.Write(plugin.Config)
+	})
+
+}
+
+// UpdateConfig updates Configurer plugin configuration in YAML format.
+// swagger:operation POST /plugin/:id/config plugin updatePluginConfig
+//
+// Update YAML configuration for Configurer plugin.
+//
+// ---
+// consumes: [application/x-yaml]
+// produces: [application/json]
+// security: [clientTokenHeader: [], clientTokenQuery: [], basicAuth: []]
+// responses:
+//   200:
+//     description: Ok
+//   400:
+//     description: Bad Request
+//     schema:
+//         $ref: "#/definitions/Error"
+//   404:
+//     description: Not Found
+//     schema:
+//         $ref: "#/definitions/Error"
+//   500:
+//     description: Internal Server Error
+//     schema:
+//         $ref: "#/definitions/Error"
+func (c *PluginAPI) UpdateConfig(ctx *gin.Context) {
+	withID(ctx, "id", func(id uint) {
+		conf := c.DB.GetPluginConfByID(id)
+		if conf == nil {
+			ctx.AbortWithError(404, errors.New("unknown plugin"))
+			return
+		}
+		if aborted := pluginOwnerOrAbort(ctx, conf); aborted {
+			return
+		}
+		instance, err := c.Manager.Instance(id)
+		if err != nil {
+			ctx.AbortWithError(404, errors.New("plugin instance not found"))
+			return
+		}
+
+		if aborted := supportOrAbort(ctx, instance, compat.Configurer); aborted {
+			return
+		}
+
+		newConf := instance.EmptyConfig()
+		newconfBytes, err := ioutil.ReadAll(ctx.Request.Body)
+		if err != nil {
+			ctx.AbortWithError(500, err)
+			return
+		}
+		if err := yaml.Unmarshal(newconfBytes, newConf); err != nil {
+			ctx.AbortWithError(400, err)
+			return
+		}
+		if err := instance.ValidateAndSetConfig(newConf); err != nil {
```

Comment:
Can we validate newConf with the schema before it will be set on the plugin?

Reply:
The schema is generated based on the struct so I don't see a possibility that the unmarshaling succeeded but the schema is incompatible.

Reply:
In should be possible to add properties like minLength to the json schema. It can be parsed successfully but doesn't have to be valid.